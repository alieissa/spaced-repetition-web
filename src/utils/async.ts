/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
