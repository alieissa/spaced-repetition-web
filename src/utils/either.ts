/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
