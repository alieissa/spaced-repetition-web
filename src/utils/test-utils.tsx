/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
