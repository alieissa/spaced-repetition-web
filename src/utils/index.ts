/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
