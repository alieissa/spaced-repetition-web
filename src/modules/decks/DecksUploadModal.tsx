/** @format */

import { ChangeEvent, useEffect, useState } from 'react'
import {
  Button,
  Loader,
  Message,
  Modal,
  ModalActions,
  ModalContent,
  ModalDescription,
  ModalHeader,
} from 'semantic-ui-react'
import 'src/App.css'
import FileSelector from 'src/components/FileSelector'
import { async } from 'src/utils'
import { ImportButton } from '../../components'
import { useUploadDecks } from './decks.hooks'
/**
 * This component contains the modal and the button that is used to toggle
 * the modal.
 */
export default function UploadDecksModal() {
  const [isModalOpen, setIsModalOpen] = useState(false)
  const [file, setFile] = useState<File | null>()
  const [uploadDecksStatus, uploadDecks, resetUploadDecks] = useUploadDecks()
  const isUploading = uploadDecksStatus.type === 'Loading'

  const handleFileSelected = (e: ChangeEvent<HTMLInputElement>) => {
    setFile(e.target.files![0])
  }

  const handleFileUpload = () => {
    uploadDecks(file!)
  }

  const handleCloseModal = () => {
    setFile(null)
    resetUploadDecks()
    setIsModalOpen(false)
  }

  useEffect(() => {
    if (uploadDecksStatus.type === 'Success') {
      handleCloseModal()
    }
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, [uploadDecksStatus.type])

  return (
    <>
      <ImportButton
        data-testid="decks-upload-modal-btn"
        onClick={() => setIsModalOpen(true)}
      />
      <Modal
        dimmer="inverted"
        size="tiny"
        open={isModalOpen}
        data-testid="decks-upload-modal"
        onClose={handleCloseModal}
      >
        <ModalHeader>Upload decks</ModalHeader>
        <ModalContent className="flex-column align-center justify-center">
          {async.match(uploadDecksStatus)({
            Untriggered: () => null,
            Loading: () => <Loader active data-testid="decks-upload-loader" />,
            Success: () => null,
            Failure: () => (
              <Message
                negative
                className="w-inherit"
                data-testid="decks-upload-error"
              >
                <Message.Header>Upload failed</Message.Header>
                <p>Please try again.</p>
              </Message>
            ),
          })}

          <ModalDescription className="pb-10">
            <p>Select file the file containing the decks</p>
          </ModalDescription>
          <div className="pb-10">{file?.name}</div>
          <FileSelector
            data-testid="decks-upload-file-input"
            onFileSelected={handleFileSelected}
          />
        </ModalContent>
        <ModalActions>
          <Button disabled={isUploading} onClick={handleCloseModal}>
            Cancel
          </Button>
          <Button
            data-testid="decks-upload-confirm-btn"
            color="green"
            disabled={!file}
            onClick={handleFileUpload}
          >
            Confirm
          </Button>
        </ModalActions>
      </Modal>
    </>
  )
}
