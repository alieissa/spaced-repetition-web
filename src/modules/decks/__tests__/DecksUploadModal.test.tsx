/** @format */

import '@testing-library/jest-dom'
import { act, screen } from '@testing-library/react'
import user from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { renderWithProviders } from 'src/utils/test-utils'
import DecksUploadModal from '../DecksUploadModal'
const mockNavigate = jest.fn(() => ({}))
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

const decksUploadUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/upload`

export const handlers = [
  rest.post(decksUploadUrl, async (__, res, ctx) => {
    return res(ctx.status(200), ctx.delay(100))
  }),
]

const uploadFile = async (inputElementId: string) => {
  const blob = new Blob(['dsfasadfdsa'])
  const file = new File([blob], 'test.json', {
    type: 'application/json',
  })

  const input = screen.getByTestId(inputElementId)
  expect(input).toBeInTheDocument()

  return user.upload(input, file)
}

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('DecksUploadModal', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<DecksUploadModal />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    it('should display loading when upload in progress', async () => {
      // Assemble
      server.use(
        rest.post(decksUploadUrl, async (__, res, ctx) => {
          return res(ctx.status(200), ctx.delay(100))
        }),
      )
      renderWithProviders(<DecksUploadModal />)
      const deckImportBtn = screen.getByTestId('decks-upload-modal-btn')
      await act(() => user.click(deckImportBtn))
      await act(() => uploadFile('decks-upload-file-input'))

      // Act
      const decksUploadConfirmBtn = screen.getByTestId(
        'decks-upload-confirm-btn',
      )
      await act(() => user.click(decksUploadConfirmBtn))

      // Assert
      const decksUploadLoader = screen.queryByTestId('decks-upload-loader')
      expect(decksUploadLoader).toBeInTheDocument()
    })

    it('should display error when upload fails', async () => {
      // Assemble
      server.use(
        rest.post(decksUploadUrl, (__, res, ctx) => res(ctx.status(422))),
      )
      renderWithProviders(<DecksUploadModal />)
      const deckImportBtn = screen.getByTestId('decks-upload-modal-btn')
      await act(() => user.click(deckImportBtn))
      await act(() => uploadFile('decks-upload-file-input'))

      // Act
      const decksUploadConfirmBtn = screen.getByTestId(
        'decks-upload-confirm-btn',
      )
      await act(() => user.click(decksUploadConfirmBtn))

      // Assert
      const decksUploadLoader = screen.queryByTestId('decks-upload-error')
      expect(decksUploadLoader).toBeInTheDocument()
    })

    // There is no success test, because that is an implementation detail
    // and is best suited for e-2-e testing
  })
})
