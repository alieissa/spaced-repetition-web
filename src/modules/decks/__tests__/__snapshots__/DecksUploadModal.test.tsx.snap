// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DecksUploadModal view should render correctly 1`] = `
<DocumentFragment>
  <button
    class="ui button"
    data-testid="decks-upload-modal-btn"
  >
    <i
      aria-hidden="true"
      class="upload icon"
    />
  </button>
</DocumentFragment>
`;
